src/wallet/wallet.cpp

```diff
@@ -1390,65 +1390,231 @@ std::optional<PaymentSource> CWallet::SelectPaymentSource(const libzcash::Paymen
-                    auto accountId = addrMetaIt->second.GetAccountId();
-                    if (accountId.has_value()) {
-                        result = AccountPaymentSource(accountId.value(), ua.GetKnownReceiverTypes());
-                    }
+                auto accountId = this->GetUnifiedAccountId(ufvkId.value());
+                if (accountId.has_value()) {
+                    result = AccountPaymentSource(accountId.value(), ua.GetKnownReceiverTypes());
                 }
             }
         }
     }, addr);
     return result;
 }
 
+bool CWallet::PaymentSourceMatchesAddress(
+        const PaymentSource& source,
+        const CTxDestination& address) const {
+    auto self = this;
+    return std::visit(match {
+        [&](const CKeyID& keyId) {
+            CTxDestination keyIdDest = keyId;
+            return address == keyIdDest;
+        },
+        [&](const CScriptID& scriptId) {
+            CTxDestination scriptIdDest = scriptId;
+            return address == scriptIdDest;
+        },
+        [&](const libzcash::SproutPaymentAddress& addr) { return false; },
+        [&](const libzcash::SaplingPaymentAddress& addr) { return false; },
+        [&](const AccountPaymentSource& acct) {
+            if (acct.UseP2PKH() || acct.UseP2SH()) {
+                std::optional<std::pair<libzcash::UFVKId, std::optional<libzcash::diversifier_index_t>>> meta;
+                std::visit(match {
+                    [&](const CNoDestination& none) { meta = std::nullopt; },
+                    [&](const auto& addr) { meta = self->GetUFVKMetadataForReceiver(addr); }
+                }, address);
+                if (meta.has_value()) {
+                    // use the coin if the account id corresponding to the UFVK is
+                    // the payment source account.
+                    return self->GetUnifiedAccountId(meta.value().first) == std::optional(acct.GetAccountId());
+                } else {
+                    // The legacy account is treated as a single pool
+                    // of funds, and is considered to include all UTXOs
+                    // prior to the advent of unified addresses.
+                    return acct.GetAccountId() == ZCASH_LEGACY_ACCOUNT;
+                }
+            }
+            return false;
+        }
+    }, source);
+}
+
+bool CWallet::PaymentSourceMatchesAddress(
+        const PaymentSource& source,
+        const libzcash::SproutPaymentAddress& a0) const {
+    return std::visit(match {
+        [&](const CKeyID& keyId) { return false; },
+        [&](const CScriptID& scriptId) { return false; },
+        [&](const libzcash::SproutPaymentAddress& a1) { return a0 == a1; },
+        [&](const libzcash::SaplingPaymentAddress& addr) { return false; },
+        [&](const AccountPaymentSource& acct) {
+            return
+                acct.GetAccountId() == ZCASH_LEGACY_ACCOUNT &&
+                acct.GetReceiverTypes().empty();
```

Comment:
```suggestion
             return false;
```
The legacy account exists solely to enable spending keys generated by legacy APIs to be bound to the mnemonic phrase; it does not mean "treat the legacy spending keys, which were distinct pools of funds, as a synthetic unified pool".